### PR TITLE
xbgpu: add command-line arguments for beamformer

### PR DESCRIPTION
**Pull request by amishatishpatel**
Much like #535, this adds a `--beamformer` command-line argument, taking properties to set via comma-separate key=value pairs.

I may have overstepped the boundaries of the PR in some way as I have an eye on NGC-929 (a la PR #537 )
* I purposefully left a `self.n_ants` attribute in `XBEngine` (used in the `_sender_loop`), and
* I updated `populate_sensors` to take the `chan-range` sensor default values as an arg.

It's a conundrum I regularly dance around (do the straightforward thing or be less abrupt with the change), please share your grievances accordingly.

Checklist (if not applicable, edit to add `(N/A)` and mark as done):

- [x] (N/A) If dependencies are added/removed: update `setup.cfg` and `.pre-commit-config.yaml`
- [x] If modules are added/removed: use sphinx-apidoc to update files in `doc/`
- [x] Ensure copyright notices are present and up-to-date
- [x] (N/A) If qualification tests are changed: attach a sample qualification report
- [x] (N/A) If design has changed: ensure documentation is up to date
- [x] (N/A) If ICD-defined sensors have been added: update `fake_servers.py` in katsdpcontroller to match

Closes NGC-905.

### CODE REVIEW

File: src/katgpucbf/xbgpu/main.py
Comment (bmerry) [inline src/katgpucbf/xbgpu/main.py:80]: I'm not seeing how channels_per_substream can be configured for an output. We receive some number of channel as input, and the same set of channels are in the output.

File: src/katgpucbf/xbgpu/output.py
Comment (bmerry) [inline src/katgpucbf/xbgpu/output.py:28]: You're also going to want a `name` field - might as well add it now (and parse it from the command line). Otherwise you're going to have to make a non-backwards-compatible change to the CLI later.

File: src/katgpucbf/xbgpu/engine.py
Comment (bmerry) [inline src/katgpucbf/xbgpu/engine.py:300]: There is a bit of a smell in the number of antennas being obtained from an Output, given that it's determined by the input. I'm not sure it belongs in the Output class at all, although I can see it may be convenient to have there for computing things like the `baselines` property. I'd suggest removing it from there for now and making everything else work (to ensure that the input side of things doesn't depend on having a X output) then think about adding it back.

**Review by bmerry — CHANGES_REQUESTED** [review]

File: src/katgpucbf/xbgpu/main.py
Comment (amishatishpatel) [reply src/katgpucbf/xbgpu/main.py:128]: Argh, my apologies, thought I got all of it (I was interrogating `args.beamformer` manually).

File: src/katgpucbf/xbgpu/output.py
Comment (amishatishpatel) [reply src/katgpucbf/xbgpu/output.py:35]: Fair points all around, it did feel a bit odd when naming/using it.

File: src/katgpucbf/xbgpu/output.py
Comment (amishatishpatel) [reply src/katgpucbf/xbgpu/output.py:48]: Yeah this sums up my thoughts more accurately. I couldn't think of any neat shorthand/nickname for the XOutput, perhaps I'll go with BOutput for consistency with XOutput.

File: src/katgpucbf/xbgpu/output.py
Comment (amishatishpatel) [reply src/katgpucbf/xbgpu/output.py:51]: Related to a few other comments, a brief discussion had during a check-in a while back, and a general consult with Andrew Martens:
* I wasn't sure where to mention my general thought process behind having a single BengineOutput (BOutput) holding details of all beams.
* Perhaps I was inception'd by the concept of the 'uber stream' you created for Narrowband, and I was convinced I need that hammer for this nail.

I promise I didn't just entirely ignore the points you raised previously. I should've have paid closer attention to a similar point discussed with AMartens around (beam config for corr2)
> the two destination multicast groups for a beam won't be contiguous and hence can't be specified as a.b.c.d+1:7148

Appreciate the way you've (re-)worded it, as it reminds me of [James's comment on NGC-929](https://skaafrica.atlassian.net/browse/NGC-929?focusedCommentId=188765)
> [...] each beam will be its own output product.

Back to the drawing board!

File: src/katgpucbf/xbgpu/output.py
Comment (amishatishpatel) [reply src/katgpucbf/xbgpu/output.py:31]: Ah, I see I see. I thought I gauged the difference as you explained and figured it was still useful/necessary here. Anyway, specifically
> Is there a need for this to be configured per Output?

I think not.

File: src/katgpucbf/xbgpu/main.py
Comment (amishatishpatel) [reply src/katgpucbf/xbgpu/main.py:80]: Again, I think this points to my internally-broken telephone where, as per [this comment on the Jira ticket](https://skaafrica.atlassian.net/browse/NGC-905?focusedCommentId=191493)
* I ended up looking at the literal output data dimensions/attributes needed to describe the output.
* In hindsight, that does seem strange, those attributes would already be part of the Engine (or even eventual Pipeline [by proxy]).

File: src/katgpucbf/xbgpu/main.py
Comment (amishatishpatel) [reply src/katgpucbf/xbgpu/main.py:346]: Definitely done by accident. As you mention, I had removed `args.dst` and relied solely what was specified in the `outputs`.
* Some (unqualified) internal strife led me to adding this back in.

Will remove now.

File: src/katgpucbf/xbgpu/engine.py
Comment (amishatishpatel) [reply src/katgpucbf/xbgpu/engine.py:300]: I'm with you. As I mentioned in [an earlier reply](https://github.com/ska-sa/katgpucbf/pull/568#discussion_r1197425815), you're entirely correct Re:
> [...] I can see it may be convenient to have there for computing things like the `baselines` property.

Shall move forward with your suggestion.

File: src/katgpucbf/xbgpu/output.py
Comment (amishatishpatel) [reply src/katgpucbf/xbgpu/output.py:39]: Like I mentioned in an earlier comment, I (eventually) took inspiration from looking at the literal attributes available/used on the output - in this case, `xsend.XSend`. Again, like you mentioned in [your weather report a bit earlier](https://github.com/ska-sa/katgpucbf/pull/568#discussion_r1196419470), it does seem to be a ~convenience thing.
* Even worse, looks like `channels` are only used to check if `channels_per_stream` are an integer multiple 🤦

File: src/katgpucbf/xbgpu/engine.py
Comment (james-smith-za) [inline src/katgpucbf/xbgpu/engine.py:479]: I don't really see the reason for this change?

**Review by james-smith-za — COMMENTED** [review]
I'm in general in support of Bruce's feedback.

I'm thinking I'd like to see something like
```
--beam=name=stephen0,dst=239.1.1.1,etc
--beam=name=stephen1,dst=239.1.1.7,etc
```
would be the incoming argument, and the number of beams being simply the number of `--beam` arguments passed.

File: test/xbgpu/test_main.py
Comment (james-smith-za) [inline test/xbgpu/test_main.py:59]: For completeness you might want to test with an invalid key as well.

File: src/katgpucbf/xbgpu/engine.py
Comment (amishatishpatel) [reply src/katgpucbf/xbgpu/engine.py:479]: In my original move to make `channels_per_substream` part of an `{X, B}Output`
* I realised that the attribute would no longer be part of the `XBEngine`, and 
* therefore wouldn't be accessible as `self.<something>` in `populate_sensors`.

I figured making `populate_sensors` accept the channel range as an arg would be a straightforward solution.
* Even setting aside the `{X, B}Output` conundrum, is it ~good practice to make an arg an attribute even if it's only used once elsewhere outside the scope of `__init__`?

Comment (james-smith-za) [reply src/katgpucbf/xbgpu/engine.py:488]: Ok, I see the logic now.

It would break from the convention of the other `populate_sensors` functions though.

Comment (amishatishpatel) [reply src/katgpucbf/xbgpu/engine.py:488]: Yeah that is the concern, maybe I'll roll it back for now and earmark it for addition/update during NGC-929.

File: src/katgpucbf/xbgpu/engine.py
Comment (james-smith-za) [inline src/katgpucbf/xbgpu/engine.py:479]: Going to need to update the docstring as well.

**Comment by amishatishpatel** [comment]
I think I've addressed all the initial comments (and rolled back/refined to the simplest version of this update). Will request another round of reviewing now.

File: src/katgpucbf/xbgpu/main.py
Comment (james-smith-za) [inline src/katgpucbf/xbgpu/main.py:67]: I know @bmerry does the F-engine OutputDicts in `main.py`, but I'm wondering whether they shouldn't be in `output.py` instead, just to keep this file a bit less cluttered.

File: src/katgpucbf/xbgpu/main.py
Comment (james-smith-za) [inline src/katgpucbf/xbgpu/main.py:287]: I am currently leaning towards suggesting removing this.

It shouldn't be possible for the beamformer to have a different number of channels per substream than the X-engine that it's co-located with.

File: src/katgpucbf/xbgpu/main.py
Comment (james-smith-za) [inline src/katgpucbf/xbgpu/main.py:314]: My sense of aesthetics almost wants the user to have to input
```
--corrprod=name=steven,dst=239.0.0.1
```
corresponding to the `--beam` argument(s). That way you bind the `dst` to the output quite nicely.

(Or some other word, I can't currently think of a better way to refer to the correlation products.)

File: src/katgpucbf/xbgpu/engine.py
Comment (james-smith-za) [inline src/katgpucbf/xbgpu/engine.py:300]: My other comment refers. I kind of suggest not including it in the `--beam=""` thing, because there should just be one value for the entire engine.

File: src/katgpucbf/xbgpu/engine.py
Comment (james-smith-za) [inline src/katgpucbf/xbgpu/engine.py:454]: @bmerry do you think it is worth protecting this somehow?

Currently `outputs` is created with the correlator output first, but that could change and this could be easy to miss. Something like
```
assert isinstance(outputs[0], XOutput)
``` 
just before the `XSend()` call.

**Review by james-smith-za — CHANGES_REQUESTED** [review]
It's getting there. Just a couple of niggles.

File: src/katgpucbf/xbgpu/main.py
Comment (bmerry) [reply src/katgpucbf/xbgpu/main.py:67]: The reason for keeping it in main.py is that it's internal to the implementation of argument parsing, rather than a public API. In fact I think it may be worth using an underscore prefix to emphasise that it's not part of the public API of main.py (in which case, do the same in fgpu.main).

File: src/katgpucbf/xbgpu/main.py
Comment (bmerry) [inline src/katgpucbf/xbgpu/main.py:74]:
```suggestion
    :class:`BOutput` for the valid keys and types. The following
```

File: src/katgpucbf/xbgpu/main.py
Comment (bmerry) [reply src/katgpucbf/xbgpu/main.py:314]: I agree with this. In fgpu I started out by keeping backwards compatibility in the first PR and later updating it to take the wideband config in `--wideband`, but doing that involved a lot of incremental refactoring and I think it's going to be cleaner to bite the bullet now.

File: src/katgpucbf/xbgpu/output.py
Comment (bmerry) [inline src/katgpucbf/xbgpu/output.py:38]: I'd suggest making the integration time (or whatever the internal representation of it is) a property of XOutput. In theory (although I can't really see a use case), you could have an engine produce two baseline-correlation-product streams with different integration times.

File: src/katgpucbf/xbgpu/main.py
Comment (bmerry) [reply src/katgpucbf/xbgpu/main.py:287]: In theory the output could be subdivided more finely than the input, and maybe someone will want that (e.g. to parallelise a beam across more receivers). But I'm not expecting that to happen, and we can add CLI support for that later if we decide to implement it. For now I'd recommend that channels-per-substream is considered an input property, not an output property.

File: src/katgpucbf/xbgpu/engine.py
Comment (bmerry) [inline src/katgpucbf/xbgpu/engine.py:301]: I wouldn't bother. If the user wants it, they can have it. In production the addresses are generated by the computer, so you don't need to protect the user from finger trouble.

File: src/katgpucbf/xbgpu/engine.py
Comment (bmerry) [reply src/katgpucbf/xbgpu/engine.py:300]: Agreed. If channels_per_substream is removed from the output, this can be reverted.

File: src/katgpucbf/xbgpu/engine.py
Comment (bmerry) [reply src/katgpucbf/xbgpu/engine.py:454]: I wouldn't worry at this point. This code will all go through major refactoring anyway.

File: test/xbgpu/test_main.py
Comment (bmerry) [inline test/xbgpu/test_main.py:63]: I'd remove the duplicate channels_per_substream, so that there is only one error with the argument.

**Review by bmerry — CHANGES_REQUESTED** [review]
Generally happy except for the inclusion of channels_per_substream in the Output class.

It occurs to me that some of the back-and-forth is driven by tension about what the Output classes are for.  In fgpu they serve several functions: they hold configuration that comes directly from the command line, but they also function as calculators for derived output properties. That works well in fgpu because the information needed for the derived properties is already in place. For xbgpu, a lot of derived properties depend on the input configuration (e.g. number of antennas or channels-per-substream), so it doesn't work so well. Once you get to implementing the beamformer you might need/want helper classes to track all the properties of an output stream, but I think it will work better to have those be distinct from the classes that hold CLI configuration (and we can bikeshed how to name things at that point).

File: src/katgpucbf/xbgpu/main.py
Comment (james-smith-za) [reply src/katgpucbf/xbgpu/main.py:67]: Ok, that makes sense.

In which case I'd suggest making that a separate, technical-debt ticket. @amishatishpatel, ignore this comment for the purposes of this PR.

File: src/katgpucbf/xbgpu/output.py
Comment (amishatishpatel) [reply src/katgpucbf/xbgpu/output.py:38]: As mentioned earlier, I'm absolutely for this. Fair warning, this comment did involve some ~overthinking on my part.

In going through implementing it, and with the context of [your review summation](https://github.com/ska-sa/katgpucbf/pull/568#pullrequestreview-1452387027), I guess I should conclude
* An Output class shall cart around input config items as long as they are of use and we can agree it's useful (?)

e.g. I did away with the `antennas` attribute on `XOutput` because the calculation of the `n_baselines` property really was for convenience. 

For the record, it depends on three input config items used for `timestamp_increment_per_accumulation`
* `heap_accumulation_threshold`
* `samples_between_spectra`, and
* `spectra_per_heap`.

That seems like a lot to me, but if it's needed then who am I to judge 🤷 I'll go ahead with 
```
class XOutput(Output):
    heap_accumulation_threshold: int
    samples_between_spectra: int
    spectra_per_heap: int

    @property
    def timestamp_increment_per_accumulation(self) -> int:
        .
        .
        .
```

Comment (bmerry) [reply src/katgpucbf/xbgpu/output.py:38]: Not quite what I had in mind. After reviewing the current CLI, what I meant was to add just `heap_accumulation_threshold` to the class (that is the "integration time", just specified in a units of heaps).

Comment (amishatishpatel) [reply src/katgpucbf/xbgpu/output.py]: Oh dear, if only I'd seen this 30 seconds ago. Apologies, rolling back now.

File: scratch/xbgpu/run-xbgpu.sh
Comment (bmerry) [inline scratch/xbgpu/run-xbgpu.sh:49]: Don't forget to remove the original `--heap-accumulation-threshold`.

File: src/katgpucbf/xbgpu/main.py
Comment (bmerry) [inline src/katgpucbf/xbgpu/main.py:56]: I'm tempted to say let's just take the opportunity to make it required for the user to pass it, rather than making an arbitrary default that's only applicable for one use case. But I also don't want to hold up this PR for longer, so maybe it should be a ticket.

File: src/katgpucbf/xbgpu/main.py
Comment (bmerry) [inline src/katgpucbf/xbgpu/main.py:162]: Perhaps this comment is no longer relevant?

File: src/katgpucbf/xbgpu/main.py
Comment (bmerry) [inline src/katgpucbf/xbgpu/main.py:148]: Doesn't seem to be required (there is a default), although see my earlier comment suggesting that we make it required.

File: src/katgpucbf/xbgpu/main.py
Comment (bmerry) [inline src/katgpucbf/xbgpu/main.py:186]:
```suggestion
        help="Add a baseline-correlation-products output (may be repeated). The required keys are: name, dst.",
```

**Review by bmerry — CHANGES_REQUESTED** [review]
Looking pretty good.

File: src/katgpucbf/xbgpu/main.py
Comment (amishatishpatel) [reply src/katgpucbf/xbgpu/main.py]: [As requested](https://skaafrica.atlassian.net/browse/NGC-988).

File: src/katgpucbf/xbgpu/main.py
Comment (bmerry) [inline src/katgpucbf/xbgpu/main.py:184]:
```suggestion
        help="Add a baseline-correlation-products output (may be repeated). The required keys are: name, dst, heap_accumulation_threshold.",
```

**Review by bmerry — CHANGES_REQUESTED** [review]
Just one place you missed documenting that heap_accumulation_threshold is now required. I'm happy apart from that, and with that fixed you can assume I accept without another round of review.

**Comment by amishatishpatel** [comment]
@james-smith-za let me know if I should wait for your review before merging.

**Review by james-smith-za — APPROVED** [review]
I like it. Some nit-picks, but they're completely stylistic. Feel free to ignore.

File: src/katgpucbf/xbgpu/main.py
Comment (bmerry) [inline src/katgpucbf/xbgpu/main.py:352]: I think this line is a leftover?

File: src/katgpucbf/xbgpu/engine.py
Comment (bmerry) [reply src/katgpucbf/xbgpu/engine.py:284]: My feeling is that assertions are fine if you believe that it can never fail, no matter what the user does, and you want to document that belief (and the computer telling you that your believe is wrong is a bonus). That way if it's eliminated by `python -O` you don't really care.

I don't think that's applicable in this case, because if the user doesn't pass `--corrprod`, this assertion will fail. So it would be a little better to replace this with an exception that explains that `--corrprod` is required. But also as James says, this is a stepping stone and we don't have to make things perfect if we know we're about to rip them out anyway.

**Review by bmerry — CHANGES_REQUESTED** [review]
By the way, don't forget that you'll need to update katsdpcontroller to pass `--corrprod`.

**Comment by amishatishpatel** [comment]
> By the way, don't forget that you'll need to update katsdpcontroller to pass `--corrprod`.

This is the kind of reminder I needed after this rollercoaster of a PR. Just kept going with the suggestions and kinda forgot the greater context of operation. As always, please and thank you!

**Pull request merged**
